pkg/scheduler/framework/plugins/noderesources/fit.go

```diff
@@ -364,19 +364,28 @@ func (f *Fit) isSchedulableAfterPodScaleDown(targetPod, originalPod, modifiedPod
 // isSchedulableAfterNodeChange is invoked whenever a node added or changed. It checks whether
 // that change made a previously unschedulable pod schedulable.
 func (f *Fit) isSchedulableAfterNodeChange(logger klog.Logger, pod *v1.Pod, oldObj, newObj interface{}) (framework.QueueingHint, error) {
-	_, modifiedNode, err := schedutil.As[*v1.Node](oldObj, newObj)
+	originalNode, modifiedNode, err := schedutil.As[*v1.Node](oldObj, newObj)
 	if err != nil {
 		return framework.Queue, err
 	}
-	// TODO: also check if the original node meets the pod's resource requestments once preCheck is completely removed.
-	// See: https://github.com/kubernetes/kubernetes/issues/110175
-	if isFit(pod, modifiedNode) {
-		logger.V(5).Info("node was updated, and may fit with the pod's resource requestments", "pod", klog.KObj(pod), "node", klog.KObj(modifiedNode))
+	// Leaving in the queue, since the pod won't fit into the modified node anyway.
+	if !isFit(pod, modifiedNode) {
+		logger.V(5).Info("node was created or updated, but it doesn't have enough resource(s) to accommodate this pod", "pod", klog.KObj(pod), "node", klog.KObj(modifiedNode))
+		return framework.QueueSkip, nil
+	}
+	// The pod will fit, so since it's add, unblock scheduling.
+	if originalNode == nil {
+		logger.V(5).Info("node was added, and may fit with the pod's resource requestments", "pod", klog.KObj(pod), "node", klog.KObj(modifiedNode))
 		return framework.Queue, nil
 	}
-
-	logger.V(5).Info("node was created or updated, but it doesn't have enough resource(s) to accommodate this pod", "pod", klog.KObj(pod), "node", klog.KObj(modifiedNode))
-	return framework.QueueSkip, nil
+	// The pod will fit, but since there was a fit in the original Node, the change won't make the pod schedulable.
+	if isFit(pod, originalNode) {
```

Comment:
Let's say `pod` requests 2 CPU, `originalNode` has 2 allocatable CPU, and `modifiedNode` has 10 allocatable CPU.
`originalNode` -> `modifiedNode` is match -> match though, we probably should still return `Queue`, right? because the Node now can accept more Pods.

I know I put the node resource fit plugin in the target list of the issue though (sorry!), maybe we cannot improve further for this?

Reply:
🤔 When `isSchedulableAfterNodeChange` is called, shouldn't we know for sure the pod is rejected by the plugin last time(if everything works fine). i.e. isFit(pod, originalNode) should always return false?

> Let's say pod requests 2 CPU, originalNode has 2 allocatable CPU, and modifiedNode has 10 allocatable CPU.

Will this pod has `noderesources` in its `UnschedulablePlugins` at the first place?

>  isFit(pod, originalNode) should always return false

If my understanding is right, I think we can simply return queue here if `isFit(pod, modifiedNode)`, we don't even need to check if this is a node/Add or node/Update event.

Reply:
> i.e. isFit(pod, originalNode) should always return false?

No, this updated Node may not be the cause putting NodeResource in pInfo.UnschedulablePlugins.
Let's say there are three Nodes. When two nodes are rejected by NodeAffinity, and the other one is rejected by NodeResource, the pod will have NodeAffinity and NodeResource in the list.
So, in such cases, the Node update might be for node-1 or node-2, not node-3 that directly caused the rejection from NodeResource. `isFit(pod, node-1)` would be true.

> Will this pod has noderesources in its UnschedulablePlugins at the first place?

Yes, if, for example, the Node has another scheduled Pod requesting two CPUs.

Reply:
Right, I got your scenario. 

BUT:
Let's say update is on node-1, since we failed schedule pod to node-1 last time. It means some plugin rejected node-1(in your case it's nodeAffinity). 
In this scenario, even if node-1 got updated, and isFit(pod, modeified-node-1) returns true, return Queue in noderesources will still make the pod fail scheduling, right? It should be the plugin which rejected the node-1 to determine whether to requeue the pod on the node-1/Update event.

So I think we can still check if the original node meets the pod's resource requestments here?

Or maybe I still haven't understand your thought. Can you list a scenario where `wasFit` is true, but we still should requeue the pod and it will success the scheduling?

Reply:
Right, my previous example was not a good one.

> Can you list a scenario where wasFit is true, but we still should requeue the pod and it will success the scheduling?

1. The Pod with 2 CPU request is about to proceed to the scheduling cycle.
2. Pod is going thru the scheduling; `node-a` is rejected by NodeResource; it has 2 allocatable CPU, but it has a running Pod requesting 2 CPU. So, 2 - 2 = 0 CPU available.
3. The Pod goes back to the queue w/ `unschedPlugin: NodeResource`
4. `Node/UpdateNodeAllocatable` is triggered and now the Node can allocate 4 CPUs. It will make the Pod schedulable. But, in the QHint, `wasFit` is true because it doesn't take a running Pod into consideration.

Note that, `isFit` checks if the Pod can run on the Node _if the Node is empty_. We don't check other running Pods on the updated Node. This is because, when we implemented QHints, we had a policy where we only refer to added/updated/deleted resource + target Pod at every QHint so that we can prevent QHint from being too complicated.

Reply:
> Note that, isFit checks if the Pod can run on the Node if the Node is empty. 

Ah, I was not aware of this 🤦.  I checked the code and yes, `Requested` of node is indeed empty in `isFit`.

> Let's say pod requests 2 CPU, originalNode has 2 allocatable CPU, and modifiedNode has 10 allocatable CPU.
originalNode -> modifiedNode is fit -> fit though, we probably should still return Queue, right?

OK, now I agree with you. It seems we can't depend on `wasFit` since we are not aware of the running Pods.

Reply:
If we discuss this, should we skip only if `!isFit(pod, modifiedNode)`

Reply:
> Note that, `isFit` checks if the Pod can run on the Node _if the Node is empty_. We don't check other running Pods on the updated Node. This is because, when we implemented QHints, we had a policy where we only refer to added/updated/deleted resource + target Pod at every QHint so that we can prevent QHint from being too complicated.

I see. Shouldn't the plugin be called with available resources then? Is there a case where total resources would matter?

Reply:
> Shouldn't the plugin be called with available resources then?

In theory, yes, but fetching the available resources of nodes would impose a greater burden on the function. I recall discussing somewhere that we aim for QHint to remain simple and reduce the requests as much as possible.

Reply:
Is it possible to check whether any of the resources specified by the Pod actually increased?

Reply:
> Is it possible to check whether any of the resources specified by the Pod actually increased?

I think that is a valid point!  cc @sanposhiho

Reply:
That's possible, yes.

Reply:
PTAL. Rewrote to unblock only when the requested resources were increased.